### DISCUSSION TITLE
Sine wave in ImageMagick?

### DISCUSSION THREAD

**Discussion by claus-01** [ImageMagick/ImageMagick#3784] [Digital Image Processing]
Hi,

I would like to know if it is possible to render a sine wave using ImageMagick? And--if so--how is this accomplished?

TIA,

Claus

**Comment by snibgo** [comment]
What version of IM, on what platform? I'll assume IM v7 on Windows BAT.

I assume you want to show a graph of a sine wave.
```
magick ^
  -size 600x1 gradient:Black-White ^
  -evaluate Sin 2 ^
  +write x.png ^
  -scale "600x600^!" ^
  -size 600x600 gradient: ^
  -compose MinusDst -composite ^
  -fill #fff +opaque #000 ^
  -morphology edgeout diamond:1 ^
  outgraph.png
```
![x2](https://user-images.githubusercontent.com/33812368/120611168-cadcee80-c44b-11eb-86f1-535da1ab0a5a.png)

This generates two cycles of a sine wave. If you want a different function, change the "-evaluate Sin 2" line.

If you want a larger size, change every "600" to whatever you want.

**Reply by claus-01** [reply]
Oh, thanks. Actually, I am using Ubuntu 20.04 & IM 7. For a Linux shell, how would I have to modify this code? Just replace ^ with /? And: What is this for: ''-size 600x1 gradient:Black-White ^? Would anti-aliasing & a broader line be possible?

Claus

**Reply by claus-01** [reply]
P. S.: I modified the script to

magick /
  -size 600x1 gradient:White-Red /
  -evaluate Sin 2 /
  +write x.png /
  -scale "625x625/!" /
  -size 625x625 gradient: /
  -compose MinusDst -composite /
  -fill #ff0000 +opaque #fff /
  -morphology edgeout diamond:3 /
  -antialias
  outgraph.png

but I only got various error messages. What would be the correct code?

Claus

**Comment by snibgo** [comment]
For Linux shell, change the line-end carets (^) to backslash (\). Not forwards slash (/). Also change any double %% to single %.

I said:
> If you want a larger size, change every "600" to whatever you want.

... but you changed some to 625, leaving others at 600. These must all be the same number.

For a thicker line, add "-morphology dilate plus:1" or similar at the end.

For different colours, change white and black to whatever you want.

For anti-aliasing, you can make the image larger (and the line thicker) then at the end resize down to whatever size you want.
```
magick ^
  -size 2400x1 gradient:Black-White ^
  -evaluate Sin 2 ^
  -scale "2400x2400^!" ^
  -size 2400x2400 gradient: ^
  -compose MinusDst -composite ^
  -fill #fff +opaque #000 ^
  -morphology edgeout diamond:1 ^
  -morphology dilate plus:8 ^
  -fill Red -opaque White ^
  -fill Blue -opaque Black ^
  -resize 25%% ^
  x2.png
```
[EDIT: I removed the superfluous but harmless "+write x.png".]

![x2](https://user-images.githubusercontent.com/33812368/120621025-6e7ecc80-c455-11eb-8857-c3a749ac3490.png)

**Reply by claus-01** [reply]
Thanks!

Claus

**Comment by claus-01** [comment]
P. S.: I made a shell scipt out of you code, but got several error messages:

 ./sine_wave.sh
magick: missing output filename `-fill' at CLI arg 17 @ error/magick-cli.c/ProcessCommandOptions/524.
./sine_wave.sh: 11: -morphology: not found

What do I have to modify?

TIA,

Claus

**Comment by snibgo** [comment]
In bash, hash # is a special character, so needs to be escaped with backslah \ or the colour needs to be quoted "#fff". This works for me, in bash:
```
magick \
  -size 2400x1 gradient:Black-White \
  -evaluate Sin 2 \
  -scale "2400x2400!" \
  -size 2400x2400 gradient: \
  -compose MinusDst -composite \
  -fill "#fff" +opaque "#000" \
  -morphology edgeout diamond:1 \
  -morphology dilate plus:8 \
  -fill Red -opaque White \
  -fill Blue -opaque Black \
  -colorspace RGB \
  -resize 25% \
  -colorspace sRGB \
  x2.png
```

**Reply by claus-01** [reply]
Now it works. Many thanks. The only thing is that the curve is slightly cut off both at the bottom and the top. How can I avoid this?

Claus

**Comment by snibgo** [comment]
The central line of the curve touches the top and the bottom.

If you want the dilated line to exactly fit _within_ the top and bottom, there are many ways of doing this. For example, change the "-evaluate" line near the top to:
```
  -evaluate Sin 2 -evaluate Multiply 0.98 -evaluate Add 1%
```
This shrinks the curve vertically to 98% of its height, and adds 1% so we get the same gap at the top and bottom. If you want an exact fit, adjust this for your dilation, or simply add "-trim" before writing the image to remove any blue rows from the top and bottom..

**Comment by snibgo** [comment]
Here is a more precise way to get the dilated line to fit within the image. The "-morphology dilate plus:8" increases the curve by 8 pixels at the top and 8 at the bottom, so we can add 8 rows to the image top and bottom before dilating. At that stage, we have a white curve on a black background, so we need to add black rows.

Windows BAT script:
```
magick ^
  -size 2400x1 gradient:Black-White ^
  -evaluate Sin 2 ^
  -scale "2400x2400^!" ^
  -size 2400x2400 gradient: ^
  -compose MinusDst -composite ^
  -fill #fff +opaque #000 ^
  -morphology edgeout diamond:1 ^
  -size 2400x8 xc:Black +swap -append ^
  -size 2400x8 xc:Black -append +repage ^
  -morphology dilate plus:8 ^
  -fill Red -opaque White ^
  -fill Blue -opaque Black ^
  -colorspace RGB ^
  -resize 25%% ^
  -colorspace sRGB ^
  x2.png
```
